aws-opentelemetry-distro/src/amazon/opentelemetry/distro/code_correlation/internal/packages.py

```diff
@@ -0,0 +1,376 @@
+from pathlib import Path
+from typing import Any, Callable, Dict, List, NamedTuple, Optional, Set, Union
+
+from ..config import AwsCodeCorrelationConfig
+
+# Module-level constants
+_logger = logging.getLogger(__name__)
+
+# Configuration
+_code_attributes_config = AwsCodeCorrelationConfig.from_env()
+
+# Global caching variables
+_sys_path_hash: Optional[int] = None
+_resolved_sys_path: List[Path] = []
+
+# Standard library paths (computed once at module load)
+_STDLIB_PATH = Path(sysconfig.get_path("stdlib")).resolve()
+_PLATSTDLIB_PATH = Path(sysconfig.get_path("platstdlib")).resolve()
+_PURELIB_PATH = Path(sysconfig.get_path("purelib")).resolve()
+_PLATLIB_PATH = Path(sysconfig.get_path("platlib")).resolve()
+
+
+class Distribution(NamedTuple):
+    """Represents a Python distribution with name and version."""
+
+    name: str
+    version: str
+
+
+def _validate_void_function(func: Callable, argspec: FullArgSpec) -> bool:
+    """Check if a function has no arguments or special characteristics."""
+    return not (
+        argspec.args
+        or argspec.varargs
+        or argspec.varkw
+        or argspec.defaults
+        or argspec.kwonlyargs
+        or argspec.kwonlydefaults
+        or isgeneratorfunction(func)
+    )
+
+
+def execute_once(func: Callable) -> Callable:
+    """
+    Decorator that ensures a function is executed only once.
+
+    Args:
+        func: Function to be decorated (must have no arguments)
+
+    Returns:
+        Wrapped function that caches its result
+
+    Raises:
+        ValueError: If the function has arguments
+    """
+    argspec = getfullargspec(func)
+    if not _validate_void_function(func, argspec):
+        raise ValueError("The execute_once decorator can only be applied to functions with no arguments")
+
+    @wraps(func)
+    def wrapper() -> Any:
+        try:
+            result, exception = func.__execute_once_result__  # type: ignore[attr-defined]
+        except AttributeError:
+            try:
+                result = func()
+                exception = None
+            except Exception as error:  # pylint: disable=broad-exception-caught
+                result = None
+                exception = error
+            func.__execute_once_result__ = result, exception  # type: ignore[attr-defined]
+
+        if exception is not None:
+            raise exception
+
+        return result
+
+    return wrapper
+
+
+def _determine_effective_root(relative_path: Path, parent_path: Path) -> str:
+    """
+    Determine the effective root module for a given path.
+
+    Args:
+        relative_path: Path relative to the parent
+        parent_path: Parent directory path
+
+    Returns:
+        Root module name
+    """
+    base_name = relative_path.parts[0]
+    root_dir = parent_path / base_name
+
+    if root_dir.is_dir() and (root_dir / "__init__.py").exists():
+        return base_name
+    return "/".join(relative_path.parts[:2])
```

Comment:
Although ADOT python doesn't support Windows yet, I'd like to ensure it is OS agnostic and works for Windows file path as well.